File: apps/runtime/src/models/catalog.rs
//! The model catalog: who publishes a model, what it costs and how much it
//! holds — 用户实测反馈 F10.
//!
//! ## Where the numbers come from
//!
//! Until now every price was a constant transcribed by hand from a vendor's
//! pricing page, and the user's first question about the cost panel was "whose
//! prices are these?" — a question the product could not answer. It can now:
//! prices and context windows come from **models.dev**, an MIT-licensed,
//! community-maintained catalog covering every provider our CLIs talk to.
//!
//! `GET https://models.dev/api.json` answers one JSON object, keyed by
//! provider id, with this shape (transcribed from the live document on
//! 2026-09-13; 213 providers, 7784 models, ~4.6 MB):
//!
//! ```json
//! {
//!   "anthropic": {
//!     "id": "anthropic",
//!     "name": "Anthropic",
//!     "doc": "https://docs.anthropic.com/…",
//!     "models": {
//!       "claude-sonnet-4-6": {
//!         "id": "claude-sonnet-4-6",
//!         "name": "Claude Sonnet 4.6",
//!         "reasoning": true,
//!         "release_date": "2026-02-17",
//!         "last_updated": "2026-03-13",
//!         "limit": { "context": 1000000, "output": 128000 },
//!         "cost": { "input": 3, "output": 15, "cache_read": 0.3, "cache_write": 3.75 }
//!       }
//!     }
//!   }
//! }
//! ```
//!
//! Costs are USD per **million** tokens, which is the unit `ModelPrice` already
//! uses. Some rows carry extra cost keys (`tiers`, `context_over_200k`,
//! `input_audio`, …) for long-context or modality surcharges; those are *not*
//! read. Billing a session at a surcharge we cannot prove applied would be the
//! same mistake as inventing a price, and rule 1 of `pricing.rs` stands:
//! a number we cannot vouch for is better left out.
//!
//! ## What is kept
//!
//! Only the providers whose models our adapters actually run (see
//! [`KEPT_PROVIDERS`]). That turns a 4.6 MB document into a ~100 KB cache and
//! keeps the parse off the hot path; a model from anywhere else stays unpriced
//! exactly as it is today, and `<data_dir>/model-pricing.json` remains the way
//! to price it by hand.
//!
//! ## When it is read
//!
//! The network is touched by the Runtime and nowhere else. The cache in
//! `<data_dir>/models-catalog.json` is loaded at startup; a cache older than
//! [`REFRESH_INTERVAL`] triggers one fetch, and another every 24 h after that.
//! A failed fetch is not an error state — the cache keeps answering, and with
//! no cache at all the built-in tables do, which is why every reader asks for
//! a fallback rather than for "the" catalog.

use std::{
    collections::BTreeMap,
    path::{Path, PathBuf},
    sync::{Arc, LazyLock, RwLock},
    time::Duration,
};

use serde::{Deserialize, Serialize};

/// The published document. Named here rather than at a call site so the one
/// address the Runtime reaches out to is greppable.
pub const CATALOG_URL: &str = "https://models.dev/api.json";

/// `<data_dir>/models-catalog.json`.
pub const CACHE_FILE: &str = "models-catalog.json";

/// Bumped when a cache written by an older Runtime can no longer be read.
pub const CACHE_VERSION: u32 = 1;

/// One fetch a day, and one at startup when the cache is older than this.
pub const REFRESH_INTERVAL: Duration = Duration::from_secs(24 * 60 * 60);

/// A response larger than this is refused before it is parsed. The live
/// document is ~4.6 MB; the ceiling is there so a redirect to something else
/// cannot make the Runtime buffer an arbitrary amount.
const MAX_RESPONSE_BYTES: u64 = 32 * 1024 * 1024;

/// Providers kept from the upstream document.
///
/// One per CLI we ship an adapter for: Claude Code (anthropic), Codex
/// (openai) and GitHub Copilot, whose catalog re-lists
/// other vendors' models under its own entitlement. A CLI that can be pointed
/// at any provider (opencode, pi, omp) is not enumerated here: its model list
/// is whatever its own configuration says, and guessing a provider for it
/// would produce a menu of models the account cannot use.
pub const KEPT_PROVIDERS: &[&str] = &["anthropic", "openai", "google", "github-copilot"];

/// USD per million tokens, as models.dev publishes them.
#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CatalogCost {
    pub input: f64,
    pub output: f64,
    #[serde(default)]
    pub cache_read: f64,
    #[serde(default)]
    pub cache_write: f64,
}

/// Token limits. Either half may be missing upstream, and a missing limit is
/// reported as missing rather than as zero.
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CatalogLimit {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub context: Option<u64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub output: Option<u64>,
}

/// One model, flattened out of the provider that publishes it.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CatalogModel {
    pub provider: String,
    pub model_id: String,
    pub name: String,
    /// Absent when the catalog publishes no usable price for this model — a
    /// subscription-only entitlement, or a row with no input/output figure.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub cost: Option<CatalogCost>,
    #[serde(default)]
    pub limit: CatalogLimit,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub release_date: Option<String>,
    #[serde(default)]
    pub reasoning: bool,
}

/// Where the catalog in memory came from. Shown in the settings page so a
/// price always has a provenance next to it.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum CatalogSource {
    /// Fetched from models.dev during this run.
    Network,
    /// Read from `<data_dir>/models-catalog.json`.
    Cache,
    /// Nothing on disk and no successful fetch: the built-in tables answer.
    BuiltIn,
}

/// The cached document, and the in-memory catalog: the same shape on disk and
/// on the wire.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Catalog {
    pub version: u32,
    pub source: CatalogSource,
    /// RFC 3339, UTC. When the models were read from models.dev — not when the
    /// cache was last loaded.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub fetched_at: Option<String>,
    pub url: String,
    pub models: Vec<CatalogModel>,
}

impl Default for Catalog {
    fn default() -> Self {
        Self {
            version: CACHE_VERSION,
            source: CatalogSource::BuiltIn,
            fetched_at: None,
            url: CATALOG_URL.to_owned(),
            models: Vec::new(),
        }
    }
}

impl Catalog {
    pub fn is_empty(&self) -> bool {
        self.models.is_empty()
    }

    /// Age of the data, not of the file. `None` when nothing was ever fetched.
    pub fn age(&self) -> Option<chrono::Duration> {
        let fetched = chrono::DateTime::parse_from_rfc3339(self.fetched_at.as_deref()?).ok()?;
        Some(chrono::Utc::now() - fetched.with_timezone(&chrono::Utc))
    }

    fn stale(&self) -> bool {
        match self.age() {
            Some(age) => {
                age > chrono::Duration::from_std(REFRESH_INTERVAL)
                    .unwrap_or(chrono::Duration::zero())
            }
            None => true,
        }
    }

    /// An exact model id, in the catalog's own spelling.
    pub fn model(&self, model_id: &str) -> Option<&CatalogModel> {
        let wanted = crate::context_models::normalize_model_id(model_id);
        self.models
            .iter()
            .find(|model| crate::context_models::normalize_model_id(&model.model_id) == wanted)
    }

    /// Every model one provider publishes, newest release first. Ties keep the
    /// catalog's order so the result is stable between fetches.
    pub fn provider_models(&self, provider: &str) -> Vec<&CatalogModel> {
        let mut models: Vec<&CatalogModel> = self
            .models
            .iter()
            .filter(|model| model.provider == provider)
            .collect();
        models.sort_by(|left, right| {
            right
                .release_date
                .as_deref()
                .unwrap_or("")
                .cmp(left.release_date.as_deref().unwrap_or(""))
        });
        models
    }
}

/* ------------------------------ upstream shape ---------------------------- */

#[derive(Deserialize)]
struct UpstreamProvider {
    #[serde(default)]
    models: BTreeMap<String, UpstreamModel>,
}

#[derive(Deserialize)]
struct UpstreamModel {
    #[serde(default)]
    id: Option<String>,
    #[serde(default)]
    name: Option<String>,
    #[serde(default)]
    reasoning: bool,
    #[serde(default)]
    release_date: Option<String>,
    #[serde(default)]
    limit: Option<UpstreamLimit>,
    #[serde(default)]
    cost: Option<UpstreamCost>,
}

#[derive(Deserialize)]
struct UpstreamLimit {
    #[serde(default)]
    context: Option<u64>,
    #[serde(default)]
    output: Option<u64>,
}

/// Only the four headline figures. Every other key upstream publishes
/// (`tiers`, `context_over_200k`, `input_audio`, `reasoning`, …) describes a
/// surcharge whose applicability we cannot observe from a transcript.
#[derive(Deserialize)]
struct UpstreamCost {
    #[serde(default)]
    input: Option<f64>,
    #[serde(default)]
    output: Option<f64>,
    #[serde(default)]
    cache_read: Option<f64>,
    #[serde(default)]
    cache_write: Option<f64>,
}

/// Parses the published document into the models we keep.
///
/// Anything that does not describe a usable row is dropped rather than
/// defaulted: a price of "missing" is a model with no price, which the cost
/// panel already knows how to show, whereas a price of zero is a lie.
pub fn parse(document: &str, fetched_at: String) -> Result<Catalog, String> {
    let upstream: BTreeMap<String, UpstreamProvider> =
        serde_json::from_str(document).map_err(|error| format!("models.dev: {error}"))?;
    let mut models = Vec::new();
    for provider in KEPT_PROVIDERS {
        let Some(entry) = upstream.get(*provider) else {
            continue;
        };
        for (key, model) in &entry.models {
            let model_id = model.id.clone().unwrap_or_else(|| key.clone());
            if model_id.trim().is_empty() {
                continue;
            }
            models.push(CatalogModel {
                provider: (*provider).to_owned(),
                name: model.name.clone().unwrap_or_else(|| model_id.clone()),
                model_id,
                cost: model.cost.as_ref().and_then(usable_cost),
                limit: CatalogLimit {
                    context: model
                        .limit
                        .as_ref()
                        .and_then(|limit| limit.context)
                        .filter(|context| *context > 0),
                    output: model
                        .limit
                        .as_ref()
                        .and_then(|limit| limit.output)
                        .filter(|output| *output > 0),
                },
                release_date: model
                    .release_date
                    .clone()
                    .filter(|date| !date.trim().is_empty()),
                reasoning: model.reasoning,
            });
        }
    }
    if models.is_empty() {
        return Err("models.dev named none of the providers we read".into());
    }
    Ok(Catalog {
        version: CACHE_VERSION,
        source: CatalogSource::Network,
        fetched_at: Some(fetched_at),
        url: CATALOG_URL.to_owned(),
        models,
    })
}

/// A cost row is usable only when both headline figures are real numbers. The
/// two cache figures default to zero, which is what "not billed separately"
/// means for a provider that does not charge for them.
fn usable_cost(cost: &UpstreamCost) -> Option<CatalogCost> {
    let finite = |value: Option<f64>| value.filter(|v| v.is_finite() && *v >= 0.0);
    Some(CatalogCost {
        input: finite(cost.input)?,
        output: finite(cost.output)?,
        cache_read: finite(cost.cache_read).unwrap_or(0.0),
        cache_write: finite(cost.cache_write).unwrap_or(0.0),
    })
}

/* --------------------------------- the cache ------------------------------ */

pub fn cache_path(data_dir: &Path) -> PathBuf {
    data_dir.join(CACHE_FILE)
}

/// Reads the cache. A missing, unreadable or future-version file is simply no
/// catalog: the built-in tables answer and a fetch is scheduled.
pub fn read_cache(path: &Path) -> Option<Catalog> {
    let contents = std::fs::read_to_string(path).ok()?;
    match serde_json::from_str::<Catalog>(&contents) {
        Ok(catalog) if catalog.version <= CACHE_VERSION && !catalog.is_empty() => Some(Catalog {
            // On disk it was whatever fetched it; in memory it is the cache.
            source: CatalogSource::Cache,
            ..catalog
        }),
        Ok(_) => None,
        Err(error) => {
            tracing::warn!(%error, path = %path.display(), "ignoring an unreadable model catalog cache");
            None
        }
    }
}

pub fn write_cache(path: &Path, catalog: &Catalog) -> std::io::Result<()> {
    let mut body = serde_json::to_string(catalog)
        .map_err(|error| std::io::Error::new(std::io::ErrorKind::InvalidData, error))?;
    body.push('\n');
    crate::hook::auth::write_private_atomically(path, body.as_bytes())
}

/* -------------------------------- the service ----------------------------- */

/// The process-wide catalog.
///
/// Global rather than a field of `AppState` because its two readers are
/// `pricing.rs` and `context_models.rs`, both of which are called from places
/// that have a model id and nothing else — a scan running on a blocking
/// thread, a context reading assembled deep inside a session. Threading a
/// handle to all of them would be a wider change than the fact being looked
/// up justifies.
static CATALOG: LazyLock<RwLock<Arc<Catalog>>> =
    LazyLock::new(|| RwLock::new(Arc::new(Catalog::default())));

/// The catalog as it stands. Never blocks on the network and never fails: an
/// empty catalog means "ask the built-in tables", which every caller does.
pub fn current() -> Arc<Catalog> {
    CATALOG
        .read()
        .unwrap_or_else(std::sync::PoisonError::into_inner)
        .clone()
}

fn install(catalog: Catalog) -> Arc<Catalog> {
    let shared = Arc::new(catalog);
    *CATALOG
        .write()
        .unwrap_or_else(std::sync::PoisonError::into_inner) = Arc::clone(&shared);
    // The per-agent menus were built from the previous catalog; a user who
    // just asked for an update should not have to wait out their TTL.
    super::agents::forget();
    shared
}

/// Loads the cache and keeps the catalog fresh: one fetch now if the cached
/// data is older than a day (or absent), then one a day.
///
/// Started from `main`. Failure is logged and nothing else: a Runtime with no
/// network still prices every model it has a built-in row for.
pub fn start(data_dir: PathBuf) {
    let path = cache_path(&data_dir);
    let cached = read_cache(&path);
    let stale = cached.as_ref().is_none_or(Catalog::stale);
    if let Some(catalog) = cached {
        tracing::info!(
            models = catalog.models.len(),
            fetched_at = catalog.fetched_at.as_deref().unwrap_or("unknown"),
            "loaded the cached model catalog"
        );
        install(catalog);
    }
    tokio::spawn(async move {
        if stale && let Err(error) = refresh(&path).await {
            tracing::warn!(%error, "could not refresh the model catalog; using what we have");
        }
        let mut ticker = tokio::time::interval(REFRESH_INTERVAL);
        ticker.tick().await;
        loop {
            ticker.tick().await;
            if let Err(error) = refresh(&path).await {
                tracing::warn!(%error, "could not refresh the model catalog; using what we have");
            }
        }
    });
}

/// Fetches models.dev once and installs the result.
///
/// The cache is only written after a successful parse, so a bad answer cannot
/// replace a good cache.
pub async fn refresh(cache: &Path) -> Result<Arc<Catalog>, String> {
    let document = fetch(CATALOG_URL).await?;
    let catalog = parse(&document, chrono::Utc::now().to_rfc3339())?;
    if let Err(error) = write_cache(cache, &catalog) {
        // A read-only data directory costs us the cache, not the catalog.
        tracing::warn!(%error, path = %cache.display(), "could not cache the model catalog");
    }
    tracing::info!(models = catalog.models.len(), "refreshed the model catalog");
    Ok(install(catalog))
}

async fn fetch(url: &str) -> Result<String, String> {
    let client = reqwest::Client::builder()
        .timeout(Duration::from_secs(30))
        .build()
        .map_err(|error| error.to_string())?;
    let response = client
        .get(url)
        .header("accept", "application/json")
        .send()
        .await
        .map_err(|error| error.to_string())?;
    if !response.status().is_success() {
        return Err(format!("{url} answered {}", response.status()));
    }
    if response
        .content_length()
        .is_some_and(|length| length > MAX_RESPONSE_BYTES)
    {
        return Err(format!(
            "{url} answered more than {MAX_RESPONSE_BYTES} bytes"
        ));
    }
    response.text().await.map_err(|error| error.to_string())
}

#[cfg(test)]
#[path = "catalog_tests.rs"]
mod tests;

File: apps/desktop/src/core/models/catalog.ts
/**
 * 模型目录：谁发布了一个模型、它多少钱、能装多少 token（用户实测反馈 F10）。
 *
 * 移植自 合并前的实现，读写**同一个**缓存文件
 * `<dataDir>/models-catalog.json`，同一份 JSON 形状。
 *
 * ## 数字从哪来
 *
 * `GET https://models.dev/api.json` answers one JSON object keyed by provider
 * id：
 *
 * ```json
 * { "anthropic": { "models": { "claude-sonnet-4-6": {
 *     "id": "claude-sonnet-4-6", "name": "Claude Sonnet 4.6",
 *     "release_date": "2026-02-17", "reasoning": true,
 *     "limit": { "context": 1000000, "output": 128000 },
 *     "cost": { "input": 3, "output": 15, "cache_read": 0.3, "cache_write": 3.75 }
 * } } } }
 * ```
 *
 * 价格是每**百万** token 的美元，和 `usage/cost.ts` 的内置表同单位。上游还发
 * 布 `tiers`、`context_over_200k`、`input_audio` 之类的附加费，**不读**：一次
 * 会话有没有触发某档附加费，从转录里看不出来，按它计费和凭空编一个价格是同一
 * 个错误。
 *
 * ## 只留四家
 *
 * 上游 213 家 provider、7784 个模型、约 4.6 MB。只留我们有适配器的那几家
 * （{@link KEPT_PROVIDERS}），缓存就是约 100 KB。别处的模型仍然「没有价格」，
 * 和今天一样。
 *
 * ## 什么时候读
 *
 * **只有 core 碰网络**。启动时读缓存；缓存比 {@link REFRESH_INTERVAL_MS} 旧就
 * 安排一次抓取，之后一天一次。抓取失败不是错误状态——缓存照常回答，没有缓存就
 * 由内置表回答，所以每个读者问的都是「现在这份目录」而不是「那份目录」。
 */

import { existsSync, readFileSync } from "node:fs";
import { join } from "node:path";

import { writeSecret } from "../paths";

/** 唯一一个对外地址，写在这里而不是调用点，方便 grep。 */
export const CATALOG_URL = "https://models.dev/api.json";

/** `<dataDir>/models-catalog.json`。 */
export const CACHE_FILE = "models-catalog.json";

/** 旧版本写的缓存读不动时 +1。 */
export const CACHE_VERSION = 1;

/** 一天一次，启动时缓存比这更旧就先抓一次。 */
export const REFRESH_INTERVAL_MS = 24 * 60 * 60 * 1000;

/**
 * 两次**用户发起**的刷新之间的最短间隔。
 *
 * 设置页的刷新按钮是一次手势，连点它不该变成连着几次 4.6 MB 的下载。冷却期内
 * 的刷新返回当前这份目录，不算失败——目录没变，只是没去问。
 */
export const REFRESH_COOLDOWN_MS = 60_000;

/**
 * 启动后第一次抓取前的等待。装配不等它，一个开了就关的 core（集成测试就是这么
 * 用的）根本不会发出这次请求。
 */
export const FIRST_REFRESH_DELAY_MS = 10_000;

/** 超过这个大小在解析前就拒绝：一次重定向不能让 core 缓冲任意多字节。 */
const MAX_RESPONSE_BYTES = 32 * 1024 * 1024;

/** 一次抓取本身的超时。 */
const FETCH_TIMEOUT_MS = 30_000;

/**
 * 从上游文档里留下的 provider。
 *
 * 一家对一个我们发了适配器的 CLI：Claude Code（anthropic）、Codex（openai）、
 * GitHub Copilot（它按自己的额度重列别家的模型）。可以指向任意 provider 的 CLI
 * （opencode、pi、omp）不在这里：它的模型是它自己配置里写的，替它猜一家
 * provider 只会列出这个账号用不了的模型。
 */
export const KEPT_PROVIDERS = [
  "anthropic",
  "openai",
  "google",
  "github-copilot",
] as const;

/** USD / 百万 token，上游怎么发布就怎么记。 */
export interface CatalogCost {
  readonly input: number;
  readonly output: number;
  readonly cacheRead: number;
  readonly cacheWrite: number;
}

/** 上游两半都可能缺，缺的报成缺，不报成 0。 */
export interface CatalogLimit {
  readonly context?: number;
  readonly output?: number;
}

export interface CatalogModel {
  readonly provider: string;
  readonly modelId: string;
  readonly name: string;
  /** 目录没公布可用价格时缺失——不是 0，0 会被读成「免费」。 */
  readonly cost?: CatalogCost;
  readonly limit: CatalogLimit;
  /** `YYYY-MM-DD`，目录里有才有。 */
  readonly releaseDate?: string;
  readonly reasoning: boolean;
}

/** 内存里这份目录从哪来。设置页把它显示在价格旁边。 */
export type CatalogSource = "network" | "cache" | "builtIn";

/** 磁盘上和内存里是同一个形状。 */
export interface Catalog {
  readonly version: number;
  readonly source: CatalogSource;
  /** RFC 3339 UTC。models.dev 被读到的时刻，不是缓存被加载的时刻。 */
  readonly fetchedAt?: string;
  readonly url: string;
  readonly models: readonly CatalogModel[];
}

export function emptyCatalog(): Catalog {
  return {
    version: CACHE_VERSION,
    source: "builtIn",
    url: CATALOG_URL,
    models: [],
  };
}

export function cachePath(dataDir: string): string {
  return join(dataDir, CACHE_FILE);
}

/** 数据的年龄（毫秒），不是文件的。从没抓到过就是 `undefined`。 */
export function ageMs(catalog: Catalog, now: number): number | undefined {
  if (catalog.fetchedAt === undefined) return undefined;
  const at = Date.parse(catalog.fetchedAt);
  return Number.isFinite(at) ? Math.max(0, now - at) : undefined;
}

/** 读不出年龄的目录一律当作过期：重抓很便宜，信一个读不懂的日期不便宜。 */
export function isStale(catalog: Catalog, now: number): boolean {
  const age = ageMs(catalog, now);
  return age === undefined || age > REFRESH_INTERVAL_MS;
}

/** 一个 provider 的全部模型，最新发布的在前。同日期保持目录自己的顺序。 */
export function providerModels(
  catalog: Catalog,
  provider: string,
): CatalogModel[] {
  return catalog.models
    .filter((model) => model.provider === provider)
    .sort((left, right) =>
      (right.releaseDate ?? "").localeCompare(left.releaseDate ?? ""),
    );
}

/* ------------------------------ 上游文档的形状 ----------------------------- */

interface UpstreamCost {
  input?: unknown;
  output?: unknown;
  cache_read?: unknown;
  cache_write?: unknown;
}

function asRecord(value: unknown): Record<string, unknown> | undefined {
  return typeof value === "object" && value !== null && !Array.isArray(value)
    ? (value as Record<string, unknown>)
    : undefined;
}

function asString(value: unknown): string | undefined {
  return typeof value === "string" && value.trim() !== ""
    ? value.trim()
    : undefined;
}

function asPositive(value: unknown): number | undefined {
  return typeof value === "number" && Number.isFinite(value) && value > 0
    ? value
    : undefined;
}

/**
 * 一行价格只有两个头条数字都是真数时才可用。两个缓存数字默认 0——对不单独为
 * 缓存计费的 provider 来说，「不单独计费」就是 0。
 */
function usableCost(raw: unknown): CatalogCost | undefined {
  const cost = asRecord(raw) as UpstreamCost | undefined;
  if (cost === undefined) return undefined;
  const finite = (value: unknown): number | undefined =>
    typeof value === "number" && Number.isFinite(value) && value >= 0
      ? value
      : undefined;
  const input = finite(cost.input);
  const output = finite(cost.output);
  if (input === undefined || output === undefined) return undefined;
  return {
    input,
    output,
    cacheRead: finite(cost.cache_read) ?? 0,
    cacheWrite: finite(cost.cache_write) ?? 0,
  };
}

/**
 * 把发布出来的文档解析成我们留下的那些模型。
 *
 * 描述不出一行可用数据的一律丢掉而不是填默认值：「没有价格」成本面板已经会显示
 * 了，而「价格是 0」是撒谎。
 */
export function parse(document: string, fetchedAt: string): Catalog {
  let upstream: unknown;
  try {
    upstream = JSON.parse(document);
  } catch (error) {
    throw new Error(
      `models.dev: ${error instanceof Error ? error.message : String(error)}`,
    );
  }
  const providers = asRecord(upstream);
  if (providers === undefined)
    throw new Error("models.dev: 不是一个 JSON 对象");

  const models: CatalogModel[] = [];
  for (const provider of KEPT_PROVIDERS) {
    const entry = asRecord(providers[provider]);
    if (entry === undefined) continue;
    const published = asRecord(entry.models);
    if (published === undefined) continue;
    for (const [key, raw] of Object.entries(published)) {
      const model = asRecord(raw);
      if (model === undefined) continue;
      const modelId = asString(model.id) ?? asString(key);
      if (modelId === undefined) continue;
      const limit = asRecord(model.limit) ?? {};
      const context = asPositive(limit.context);
      const output = asPositive(limit.output);
      const cost = usableCost(model.cost);
      const releaseDate = asString(model.release_date);
      models.push({
        provider,
        modelId,
        name: asString(model.name) ?? modelId,
        ...(cost === undefined ? {} : { cost }),
        limit: {
          ...(context === undefined ? {} : { context }),
          ...(output === undefined ? {} : { output }),
        },
        ...(releaseDate === undefined ? {} : { releaseDate }),
        reasoning: model.reasoning === true,
      });
    }
  }
  if (models.length === 0) {
    throw new Error("models.dev: 里面没有我们读的那几家 provider");
  }
  return {
    version: CACHE_VERSION,
    source: "network",
    fetchedAt,
    url: CATALOG_URL,
    models,
  };
}

/* --------------------------------- 缓存文件 -------------------------------- */

/**
 * 读缓存。文件不在、读不动、或者是更新版本写的，一律当作「没有目录」：内置表
 * 回答，同时安排一次抓取。
 */
export function readCache(path: string): Catalog | undefined {
  if (!existsSync(path)) return undefined;
  let parsed: unknown;
  try {
    parsed = JSON.parse(readFileSync(path, "utf8"));
  } catch {
    return undefined;
  }
  const document = asRecord(parsed);
  if (document === undefined) return undefined;
  const version = typeof document.version === "number" ? document.version : 0;
  if (version > CACHE_VERSION) return undefined;
  if (!Array.isArray(document.models) || document.models.length === 0) {
    return undefined;
  }
  const fetchedAt = asString(document.fetchedAt);
  return {
    version,
    // 磁盘上它是谁抓的就写的谁；在内存里它是缓存。
    source: "cache",
    ...(fetchedAt === undefined ? {} : { fetchedAt }),
    url: asString(document.url) ?? CATALOG_URL,
    models: document.models as CatalogModel[],
  };
}

export function writeCache(path: string, catalog: Catalog): void {
  writeSecret(path, `${JSON.stringify(catalog)}\n`);
}

/* ---------------------------------- 抓取 ---------------------------------- */

/** 注入点：测试给一个假的，生产用 `globalThis.fetch`。 */
export type CatalogFetch = (url: string) => Promise<string>;

export async function fetchCatalog(url: string): Promise<string> {
  const answer = await globalThis.fetch(url, {
    headers: { accept: "application/json" },
    signal: AbortSignal.timeout(FETCH_TIMEOUT_MS),
  });
  if (!answer.ok) throw new Error(`${url} 答了 ${answer.status}`);
  const length = Number(answer.headers.get("content-length") ?? "0");
  if (Number.isFinite(length) && length > MAX_RESPONSE_BYTES) {
    throw new Error(`${url} 答了超过 ${MAX_RESPONSE_BYTES} 字节`);
  }
  return answer.text();
}
